plutus-tx-plugin/src/PlutusTx/Compiler/Expr.hs

```diff
@@ -430,23 +432,23 @@ hoistExpr var t = do
                 (PIR.Def var' (PIR.mkVar () var', PIR.Strict))
                 mempty
 
-            CompileContext {ccOpts=compileOpts} <- ask
-            t' <-
-                if coProfile compileOpts==All then do
-                    let ty = PLC._varDeclType var'
-                        varName = PLC._varDeclName var'
-                    t'' <- compileExpr t
-                    thunk <- PLC.freshName "thunk"
-                    pure $
-                        traceInside varName thunk t'' ty
-                else compileExpr t
-
+            t' <- maybeProfileRhs var' =<< compileExpr t
             -- See Note [Non-strict let-bindings]
             let strict = PIR.isPure (const PIR.NonStrict) t'
 
             PIR.modifyTermDef lexName (const $ PIR.Def var' (t', if strict then PIR.Strict else PIR.NonStrict))
             pure $ PIR.mkVar () var'
 
+maybeProfileRhs :: CompilingDefault uni fun m => PLCVar uni fun -> PIRTerm uni fun -> m (PIRTerm uni fun)
+maybeProfileRhs var t = do
+    CompileContext {ccOpts=compileOpts} <- ask
+    thunk <- PLC.freshName "thunk"
+    let ty = PLC._varDeclType var
+        varName = PLC._varDeclName var
+        isFunctionOrAbstraction = case ty of { PLC.TyFun{} -> True; PLC.TyForall{} -> True; _ -> False }
+    -- Trace only if profiling is on *and* the thing being defined is a function
+    pure $ if coProfile compileOpts==All && isFunctionOrAbstraction then traceInside varName thunk t ty else t
```

Comment:
This may have better performance? Because this way it goes to `compileExpr t` right away when profiling is off.

```suggestion
            CompileContext {ccOpts=compileOpts} <- ask
            t' <- 
              if coProfile compileOpts==All then 
                maybeProfileRhs var' =<< compileExpr t
              else compileExpr t
              
            -- See Note [Non-strict let-bindings]
            let strict = PIR.isPure (const PIR.NonStrict) t'

            PIR.modifyTermDef lexName (const $ PIR.Def var' (t', if strict then PIR.Strict else PIR.NonStrict))
            pure $ PIR.mkVar () var'

maybeProfileRhs :: CompilingDefault uni fun m => PLCVar uni fun -> PIRTerm uni fun -> m (PIRTerm uni fun)
maybeProfileRhs var t = do
    thunk <- PLC.freshName "thunk"
    let ty = PLC._varDeclType var
        varName = PLC._varDeclName var
        isFunctionOrAbstraction = case ty of { PLC.TyFun{} -> True; PLC.TyForall{} -> True; _ -> False }
    -- Trace only if profiling is on *and* the thing being defined is a function
    pure $ if isFunctionOrAbstraction then traceInside varName thunk t ty else t
```

Reply:
Eh, I'm not too worried. The extra work is either lazy, or just checking the head of the type variable which should be super cheap. It *does* unconditionally allocate a unique for the "thunk", which arguably would be nice to avoid if we don't need it, I'll fix that.